services/report/raw_upload_processor.py

```diff
@@ -159,6 +186,100 @@ class SessionAdjustmentResult(object):
     partially_deleted_sessions: set
 
 
+# RUSTIFYME
+@sentry_sdk.trace
+def make_sure_orginal_report_is_using_label_ids(original_report: Report) -> bool:
+    """Makes sure that the original_report (that was pulled from DB)
+    has CoverageDatapoints that encode label_ids and not actual labels.
+    """
+    # Always point the special label to index 0
+    reverse_index_cache = {
+        SpecialLabelsEnum.CODECOV_ALL_LABELS_PLACEHOLDER.corresponding_label: 0
+    }
+    if 0 not in original_report._labels_index:
+        original_report._labels_index[
+            0
+        ] = SpecialLabelsEnum.CODECOV_ALL_LABELS_PLACEHOLDER.corresponding_label
+
+    def possibly_translate_label(label_or_id: typing.Union[str, int]) -> int:
+        if type(label_or_id) == int:
+            return label_or_id
+        if label_or_id in reverse_index_cache:
+            return reverse_index_cache[label_or_id]
+        # Search for label in the report index
+        for idx, label in original_report._labels_index.items():
+            if label == label_or_id:
+                reverse_index_cache[label] = idx
+                return idx
+        # Label is not present. Add to index.
+        # Notice that this never picks index 0, that is reserved for the special label
+        new_index = max(original_report._labels_index.keys()) + 1
+        reverse_index_cache[label_or_id] = new_index
+        # It's OK to update this here because it's inside the
+        # UploadProcessing lock, so it's exclusive access
+        original_report._labels_index[new_index] = label_or_id
```

Comment:
`_labels_index` is just an instance var, right? what would the race be?

is there an alternative to this that would not make parallel upload processing harder?

Reply:
The race would be different instances of the same report (that all pull the labels_index from the same file) writing different things to it and then trying to save the report to the same destination. But because all changes happen only in the _merge_ portion of processing it's fine. THe merge needs to be in a lock anyway.

However I should be more careful that the report might be processing and not processing at the same time... and only save it into storage if it was modified.

Anyway, it doesn't really matter for parallel processing, because the uploads will be processed into temporary reports with their indexes. You can even merge them into a single report to be merged with the original in parallel, just take care of the index (with the aux functions in this file). But ultimately you will have to put a lock in the original report when merging to it, right?